adding port v1->v2 TEST-0048 (#13)
ScreaMy7/owasp-mastg: demos/android/MASVS-RESILIENCE/MASTG-DEMO-0x4B/MASTG-DEMO-0x4B.md

Reviewer: The content does not align with the guidelines. Please refer to the guidelines and update them accordingly.https://github.com/OWASP/mastg/blob/master/.github/instructions/mastg-demo.instructions.md

---
platform: android

> title: Bypass of Xposed/LSPosed Detection via PackageManager, Method-Modifier, Maps and Stack-Trace Hooks

Reviewer: ```suggestion
title: Bypassing Xposed/LSPosed Detection via API Hooking
```

id: MASTG-DEMO-0x4B
code: [kotlin]
test: MASTG-TEST-0x49
tools: [MASTG-TOOL-0031]
kind: fail
---

## Sample

> This demo defeats four Xposed/LSPosed detection checks with a Frida script that hooks the Java APIs each routine depends on (`getPackageInfo`, `Method.getModifiers`, `BufferedReader.readLine`, `Throwable.getStackTrace`, `File.listFiles`).

Reviewer: ```suggestion
This sample uses the same code as @MASTG-DEMO-0x4A, which implements four independent Xposed/LSPosed detection routines: a `PackageManager` presence probe for known Manager package ids, a `Modifier.isNative` tripwire on guaranteed-native methods, a `/proc/self/maps` scan for foreign APK/DEX mappings, and a stack-trace and `/proc/self/task` walk for framework and native worker thread needles. This demo demonstrates bypassing all four checks by hooking the Java APIs each routine depends on (`PackageManager.getPackageInfo`, `Method.getModifiers`, `BufferedReader.readLine`, `Throwable.getStackTrace`, `File.listFiles`) so that Manager packages disappear, the `NATIVE` modifier bit is re-set, `/proc/self/maps` and `/proc/self/task` lines are filtered, and framework frames are stripped from every stack trace.

!!! note
    This is a series of correlated tests.

    - @MASTG-DEMO-0x4A is a successful test (successful defense/failed attack) against an Xposed/LSPosed instrumentation attack.
    - This test is a failed test (failed defence/successful attack) against the defenses of @MASTG-DEMO-0x4A by using a more "complex" attack.
```

Reviewer: Reference : https://mas.owasp.org/MASTG/demos/android/MASVS-RESILIENCE/MASTG-DEMO-0108/MASTG-DEMO-0108/

> {{ ../MASTG-DEMO-0x4A/MastgTest.kt # script.js }}

Reviewer: ```suggestion
{{ ../MASTG-DEMO-0x4A/MastgTest.kt }}
```

## Steps

1. Install the app on a device where the Xposed/LSPosed framework is active and at least one module is scoped to `org.owasp.mastestapp`.

> 2. Spawn the app with the Frida bypass attached.

Reviewer: ```suggestion
1. Install the app on a device (@MASTG-TECH-0005) where the Xposed/LSPosed framework is active and at least one module is scoped to `org.owasp.mastestapp`.
2. Make sure you have @MASTG-TOOL-0031 installed on your machine and the frida-server running on the device.
3. Run `run.sh` to spawn the app with the bypass script.
4. Click the **Start** button.
5. Stop the script by pressing `Ctrl+C` and/or `q` to quit the Frida CLI.

{{ script.js # run.sh }}
```

Reviewer: Reference : https://mas.owasp.org/MASTG/demos/android/MASVS-RESILIENCE/MASTG-DEMO-0108/MASTG-DEMO-0108/

> {{ run.sh }}

Reviewer: ```suggestion
```

## Observation

> The Frida console shows every hook firing while the app reports **PASS** for all four detection checks.

Reviewer: ```suggestion
The output contains the trace lines emitted by each hook as it intercepts a detection probe: five `NameNotFoundException` fakes for the known Xposed/LSPosed Manager package ids, one `Method.getModifiers` interception re-setting the `NATIVE` tripwire, three `/proc/self/maps` lines dropped from a foreign APK mapping, two stack traces with framework frames stripped, and two `/proc/self/task` entries hidden (`gmain`, `pool-frida`).
```
